doc/OutputFilter.3
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man 5.01 (Pod::Simple 3.45)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "STDIN 1"
.TH STDIN 1 2024-01-04 "perl v5.38.2" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
OutputFilter \- Enable post processing of output without fork
.SH SYNOPSIS
.IX Header "SYNOPSIS"
.Vb 1
\& use Text::OutputFilter;
\&
\& my $bucket = "";
\& tie *STDOUT, "Text::OutputFilter";
\& tie *HANDLE, "Text::OutputFilter", 4;
\& tie *HANDLE, "Text::OutputFilter", 4,  *STDOUT;
\& tie *STDOUT, "Text::OutputFilter", 4, \e$bucket;
\& tie *OUTPUT, "Text::OutputFilter", 4,  *STDOUT, sub { "$_[0]" };
.Ve
.SH DESCRIPTION
.IX Header "DESCRIPTION"
This interface enables some post-processing on output streams,
like adding a left margin.
.PP
The tied filehandle is opened unbuffered, but the output is line
buffered. The \f(CW\*(C`tie\*(C'\fR takes three optional arguments:
.IP "Left Margin" 4
.IX Item "Left Margin"
The left margin must be a positive integer and defaults to \f(CW4\fR spaces.
.IP "Output Stream" 4
.IX Item "Output Stream"
The output stream must be an already open stream, with writing
enabled. The default is \f(CW*STDOUT\fR. All input methods on the new
stream are disabled. If a reference to a scalar is passed, it will
be opened as PerlIO::scalar \- in-memory IO, scalar IO. No checks
performed to see if your perl supports it. If you want it, and your
perl does not, upgrade.
.Sp
Using \f(CW\*(C`binmode ()\*(C'\fR on the new stream is allowed and supported.
.Sp
OPEN, SEEK, and WRITE are not (yet) implemented.
.IP "Line Modifying Function" 4
.IX Item "Line Modifying Function"
The output is line buffered, to enable line-modifier functions.
The line (without newline) is passed as the only argument to the
sub-ref, whose output is printed after the prefix from the first
argument. A newline is printed after the sub-ref's output.
.Sp
To \fBfilter\fR a line, as in \fIremove\fR it from the stream, make the
sub return \fIundef\fR.
.SH TODO
.IX Header "TODO"
Tests, tests, tests.
Tests with older perls
.SH AUTHOR
.IX Header "AUTHOR"
H.Merijn Brand <h.m.brand@procura.nl>
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright (C) 2006\-2024 H.Merijn Brand for PROCURA B.V.
.PP
This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBperl\fR\|(1), \fBperlopen\fR\|(1), 'open STDOUT, "|\-"', Text::Filter


Changes
0.26	- 04 Jan 2024, H.Merijn Brand
    * Fix static doc for groff-1.24 fail
    * It is 2024

0.25	- 06 Jan 2023, H.Merijn Brand
    * It is 2023

0.24	- 01 Jan 2022, H.Merijn Brand
    * It is 2021 - No changes in 2021
    * It is 2022

0.23	- 21 Dec 2020, H.Merijn Brand
    * Fix META issue for bugtracker

0.22	- 09 Apr 2020, H.Merijn Brand
    * Provide cpanfile
    * It is 2020

0.21	- 23 Apr 2019, H.Merijn Brand
    * It is 2017
    * It is 2018
    * It is 2019
    * Make PRINT return as print would (issue #1)

0.20	- 12 May 2016, H.Merijn Brand
    * Add CONTRIBUTING.md
    * Do not print undefined lines on close
    * It is 2016
    * Tested with pperl 5.24.0

0.19	- 11 Feb 2015, H.Merijn Brand
    * Upped copyright to 2015
    * Move repo to github
    * Remove perl recommendation from META as it breaks cpan clients

0.18	- 07 Feb 2014, H.Merijn Brand
    * Upped copyright to 2012
    * Upped copyright to 2013
    * Upped copyright to 2014
    * Updated META specs to Lancaster consensus (administrative release)

0.17	- 07 Sep 2011, H.Merijn Brand
    * NAME / DISTNAME in Makefile.PL

0.16	- 07 Sep 2011, H.Merijn Brand
    * Spell checks
    * More cross-checks for META data
    * Upped copyright to 2011

0.15	- 16 Mar 2010, H.Merijn Brand
    * Added Test::NoWarnings
    * Dropped YAML spec to 1.0
    * Upped copyright to 2010

0.14	- 07 Jan 2009, H.Merijn Brand
    * Copyright 2009
    * Made the license as in perl. Was still accidentally restricted
    * Corrected e-mail
    * Generated and checked META.yml

0.13	- 02 Jan 2008, H.Merijn Brand
    * Spellchecks
    * Removed prototypes
    * Now under git
    * Added simplistic example
    * Added prereq's to Makefile.PL
    * Upped copyright year

0.12	- 02 May 2007, H.Merijn Brand
    * Pod checks were wrong (yank/put problem)
    * Added test_cover make target
    * Added Text::Filter to SEE ALSO
    * Added tests for printf and binmode
    * Fixed binmode
    * Added tests for tell, eof, and fileno
    * Added tests for input (forbidden) and NYI functionality

0.11	- 21 Mar 2007, H.Merijn Brand
    * Initial release to CPAN


OutputFilter.pm
package Text::OutputFilter;

use strict;
use warnings;

our $VERSION = "0.26";

=head1 NAME

OutputFilter - Enable post processing of output without fork

=head1 SYNOPSIS

 use Text::OutputFilter;

 my $bucket = "";
 tie *STDOUT, "Text::OutputFilter";
 tie *HANDLE, "Text::OutputFilter", 4;
 tie *HANDLE, "Text::OutputFilter", 4,  *STDOUT;
 tie *STDOUT, "Text::OutputFilter", 4, \$bucket;
 tie *OUTPUT, "Text::OutputFilter", 4,  *STDOUT, sub { "$_[0]" };

=head1 DESCRIPTION

This interface enables some post-processing on output streams,
like adding a left margin.

The tied filehandle is opened unbuffered, but the output is line
buffered. The C<tie> takes three optional arguments:

=over 4

=item Left Margin

The left margin must be a positive integer and defaults to C<4> spaces.

=item Output Stream

The output stream must be an already open stream, with writing
enabled. The default is C<*STDOUT>. All input methods on the new
stream are disabled. If a reference to a scalar is passed, it will
be opened as PerlIO::scalar - in-memory IO, scalar IO. No checks
performed to see if your perl supports it. If you want it, and your
perl does not, upgrade.

Using C<binmode ()> on the new stream is allowed and supported.

OPEN, SEEK, and WRITE are not (yet) implemented.

=item Line Modifying Function

The output is line buffered, to enable line-modifier functions.
The line (without newline) is passed as the only argument to the
sub-ref, whose output is printed after the prefix from the first
argument. A newline is printed after the sub-ref's output.

To B<filter> a line, as in I<remove> it from the stream, make the
sub return I<undef>.

=back

=head1 TODO

Tests, tests, tests.
Tests with older perls

=head1 AUTHOR

H.Merijn Brand <h.m.brand@procura.nl>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006-2024 H.Merijn Brand for PROCURA B.V.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=head1 SEE ALSO

perl(1), perlopen(1), 'open STDOUT, "|-"', Text::Filter

=cut

use Carp;

sub TIEHANDLE {
    my ($class, $lm, $io, $ref, $fno) = @_;

    defined $lm  or $lm  = 4;
    defined $io  or $io  = *STDOUT;
    defined $ref or $ref = sub { shift };

    ref $lm || $lm !~ m/^\d+$/	and
	croak "OutputFilter tie's 1st arg must be numeric";
    ref $ref eq "CODE"		or
	croak "OutputFilter tie's 3rd arg must be CODE-ref";

    my $fh;
    if (ref $io eq "GLOB" and ref *{$io}{IO} eq "IO::Handle") {
	open $fh, ">&", *{$io}{IO};
	}
    elsif (ref $io eq "SCALAR") {
	open $fh, ">", $io;
	}
    else {
	eval { $fno = fileno $io };
	defined $fno && $fno >= 0 or
	    croak "OutputFilter tie's 2nd arg must be the output handle\n";
	open $fh, ">&", $fno;
	}
    $fh or croak "OutputFilter cannot dup the output handle: $!";
    select ((select ($fh), $| = 1)[0]);

    bless {
        pfx	=> " " x $lm,
        sb	=> $ref,
        io	=> $fh,

        line	=> "",

        closed	=> 0,
        }, $class;
    } # TIEHANDLE

sub BINMODE {
    my $self = shift;
    $self->{closed} and croak "Cannot set binmode on closed filehandle";
    if (@_) {
	my $mode = shift;
	binmode $self->{io}, $mode;
	}
    else {
	binmode $self->{io};
	}
    } # BINMODE

sub FILENO {
    my $self = shift;
    fileno $self->{io};
    } # FILENO

sub _Filter_ {
    my ($nl, $pfx, $sub, $line) = @_;
    my $l = $sub->($line);
    defined $l ? $pfx . $l . ($nl ? "\n" : "") : "";
    } # _Filter_

sub PRINT {
    my $self = shift;
    my ($pfx, $io, $sub) = @{$self}{qw( pfx io sb )};

    $self->{closed} and croak "Cannot print to closed filehandle";

    my $fsep = defined $, ? $, : "";
    my $rsep = defined $\ ? $\ : "";
    my $line = $self->{line} . (join $fsep => @_) . $rsep;
    my @line = split m/\n/, $line, -1;
    $self->{line} = pop @line;
    print { $io } map { _Filter_ (1, $pfx, $sub, $_) } @line;
    } # PRINT

sub PRINTF {
    my $self = shift;
    my ($pfx, $io, $sub) = @{$self}{qw( pfx io sb )};

    # Do not delegate this to PRINT, so we can prevent sprintf side effects
    $self->{closed} and croak "Cannot print to closed filehandle";

    my $fmt = shift;
    $self->PRINT (sprintf $fmt, @_);
    } # PRINTF

sub TELL {
    my $self = shift;
    $self->{closed} and croak "Cannot tell from a closed filehandle";
    tell $self->{io};
    } # TELL

sub EOF {
    my $self = shift;
    $self->{closed};
    } # EOF

sub CLOSE {
    my $self = shift;
    my ($pfx, $io, $sub, $line) = @{$self}{qw( pfx io sb line )};
    defined $line && $line ne "" and
	print { $io } _Filter_ (0, $pfx, $sub, $line);
    $self->{closed} or close $io;
    $self->{line} = "";
    $self->{closed} = 1;
    } # CLOSE

sub UNTIE {
    my $self = shift;
    $self->{closed} or $self->CLOSE;
    $self;
    } # UNTIE

sub DESTROY {
    my $self = shift;
    $self->{closed} or $self->CLOSE;
    %$self = ();
    undef $self;
    } # DESTROY

### ###########################################################################

sub _outputOnly {
    my $name = shift;
    sub { croak "No support for $name method: File is output only" };
    } # _outputOnly

*read		= _outputOnly ("read");
*READ		= _outputOnly ("READ");
*readline	= _outputOnly ("readline");
*READLINE	= _outputOnly ("READLINE");
*getc		= _outputOnly ("getc");
*GETC		= _outputOnly ("GETC");

sub _NYI {
    my $name = shift;
    sub { croak "Support for $name method NYI" };
    } # _NYI

*open		= _NYI ("open");
*OPEN		= _NYI ("OPEN");
*seek		= _NYI ("seek");
*SEEK		= _NYI ("SEEK");
*write		= _NYI ("write");
*WRITE		= _NYI ("WRITE");

=begin comment

We do not want to document these:

=over 4

=item getc

=item open

=item read

=item readline

=item seek

=item write

=back

=end comment

=cut

1;
